File: remove_duplicates_from_sorted_array.py
'''
Given a sorted array, remove the duplicates in place such that each element appear only once and return the new length.

Do not allocate extra space for another array, you must do this in place with constant memory.

For example,
Given input array A = [1,1,2],

Your function should return length = 2, and A is now [1,2].
'''

'''
Use two pointers. Quite straightforward.
'''

class Solution:
    # @param a list of integers
    # @return an integer
    def removeDuplicates(self, A):
        if len(A) < 2:
            return len(A)
        p1 = 0
        p2 = 1
        while p2 < len(A):
            while p2 < len(A) and A[p1] == A[p2]:
                p2 += 1
            p1 += 1
            if p2 < len(A):
                A[p1] = A[p2]
        return p1

if __name__ == '__main__':
    s = Solution()
    A = [1, 1, 2, 2, 3]
    print s.removeDuplicates(A)
    print A
    A = [1, 1]
    print s.removeDuplicates(A)
    print A

File: c++/remove_duplicates_from_sorted_array.cpp
/*
Given a sorted array, remove the duplicates in place such that each element appear only once and return the new length.

Do not allocate extra space for another array, you must do this in place with constant memory.

For example,
Given input array nums = [1,1,2],

Your function should return length = 2, with the first two elements of nums being 1 and 2 respectively. It doesn't matter what you leave beyond the new length.
*/

/**
 * Two pointers. Straightforward
 */

#include "helper.h"

class Solution {
public:
    int removeDuplicates(vector<int> &nums) {
        if (nums.size() == 0) {
            return 0;
        }
        int j = 0;
        for (int i = 1; i < nums.size(); ++i) {
            if (nums[i] != nums[j]) {
                nums[++j] = nums[i];
            }
        }
        return j + 1;
    }
};

int main() {
    Solution s;
    const int arr[] = {1, 1, 2, 2, 3};
    vector<int> v(arr, arr + sizeof(arr) / sizeof(arr[0]));
    cout << s.removeDuplicates(v) << endl;
    cout << v << endl;
    return 0;
}
